File: clients/uniapp/src/features/workflow/workflow-todo-client.ts
import {
  isWorkflowTodoDetail,
  readActWorkflowTodoRequest,
  readWorkflowInstanceResponse,
  readWorkflowListMyTodosResponse,
  type ActWorkflowTodoRequest,
  type WorkflowInstanceResponse,
  type WorkflowTodoDetail,
  type WorkflowTodoResponse
} from '@fullnet/client-contracts';
import type { HttpClient } from '../../api/http';
import type { WorkflowSchemaCache } from './workflow-schema-cache';

export interface WorkflowTodoClient {
  listMine(): Promise<readonly WorkflowTodoResponse[]>;
  get(todoId: string): Promise<WorkflowTodoDetail>;
  approve(todoId: string, request: ActWorkflowTodoRequest): Promise<WorkflowInstanceResponse>;
  reject(todoId: string, request: ActWorkflowTodoRequest): Promise<WorkflowInstanceResponse>;
}

const guidPattern = /^[0-9a-f]{8}-[0-9a-f]{4}-[1-8][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i;

/** 创建只消费本人待办 API 的移动端客户端；所有响应在进入状态层前执行运行时守卫。 */
export function createWorkflowTodoClient(
  http: HttpClient,
  schemaCache: WorkflowSchemaCache
): WorkflowTodoClient {
  return {
    async listMine() {
      const value = await http.request<unknown>({
        path: '/api/v1/workflow/todos/mine'
      });
      return readWorkflowListMyTodosResponse(value);
    },
    async get(todoId) {
      const normalizedTodoId = requireTodoId(todoId);
      const value = await http.request<unknown>({
        path: `/api/v1/workflow/todos/${normalizedTodoId}/runtime`
      });
      if (!isWorkflowTodoDetail(value)) {
        throw new TypeError('workflow.todo.invalid-detail');
      }

      const cachedSchema = schemaCache.read(value.formVersionId, value.formSchemaHash);
      if (cachedSchema) {
        return { ...value, formSchema: cachedSchema };
      }

      schemaCache.write(value.formVersionId, value.formSchemaHash, value.formSchema);
      return value;
    },
    approve(todoId, request) {
      return act(http, todoId, 'approve', request);
    },
    reject(todoId, request) {
      return act(http, todoId, 'reject', request);
    }
  };
}

async function act(
  http: HttpClient,
  todoId: string,
  action: 'approve' | 'reject',
  request: ActWorkflowTodoRequest
): Promise<WorkflowInstanceResponse> {
  const normalizedTodoId = requireTodoId(todoId);
  const body = readActWorkflowTodoRequest(request);
  const value = await http.request<unknown>({
    path: `/api/v1/workflow/todos/${normalizedTodoId}/${action}`,
    method: 'POST',
    data: body
  });
  return readWorkflowInstanceResponse(value);
}

function requireTodoId(value: string): string {
  if (!guidPattern.test(value)) {
    throw new TypeError('workflow.todo.invalid-id');
  }
  return value.toLowerCase();
}

File: clients/flutter/lib/features/workflow/workflow_todo_client.dart
import '../../core/http/full_net_http_client.dart';
import 'workflow_models.dart';

/// 消费工作流待办 OpenAPI；所有响应在进入 UI 前执行运行时守卫。
class WorkflowTodoClient {
  WorkflowTodoClient(this._http);

  final FullNetHttpClient _http;

  Future<List<WorkflowTodoListItem>> listMine({
    int page = 1,
    int pageSize = 20,
  }) async {
    final value = await _http.getJson(
      '/api/v1/workflow/todos/mine?page=$page&pageSize=$pageSize',
    );
    return parseWorkflowTodoListPage(value);
  }

  Future<WorkflowTodoDetail> getRuntime(String todoId) async {
    final value = await _http.getJson('/api/v1/workflow/todos/$todoId/runtime');
    if (value is! Map<String, dynamic>) {
      throw const FormatException('WorkflowTodoRuntimeResponse must be an object');
    }
    return WorkflowTodoDetail.fromJson(value);
  }

  Future<void> approve(String todoId, ActWorkflowTodoRequest request) =>
      _act(todoId, 'approve', request);

  Future<void> reject(String todoId, ActWorkflowTodoRequest request) =>
      _act(todoId, 'reject', request);

  Future<void> _act(
    String todoId,
    String action,
    ActWorkflowTodoRequest request,
  ) async {
    await _http.postJson(
      '/api/v1/workflow/todos/$todoId/$action',
      body: request.toJson(),
    );
  }
}
